src/flow/codeGeneration.js

```diff
@@ -141,104 +136,102 @@ export function typeDeclarationForOperation(
     variables,
     fields,
     fragmentSpreads,
-    inlineFragments,
     fragmentsReferenced,
     source,
-  },
-  fragmentsWithTypenameField
-) {
-  if (hasTypenameField(fields)) {
-    console.error('__typename on operations are not yet supported');
   }
+) {
   const interfaceName = interfaceNameFromOperation({operationName, operationType});
-  const properties = propertiesFromFields(generator.context, fields, {
-    typeNameSuffix: `From${operationName}`
-  });
-
-  typeDeclaration(generator, { interfaceName }, () => {
+  const properties = propertiesFromFields(generator.context, fields);
+  typeDeclaration(generator, {
+    interfaceName,
+  }, () => {
     propertyDeclarations(generator, properties, true);
   });
-
-  properties.forEach(({ fragmentSpreads, inlineFragments, bareTypeName }) => {
-    if (fragmentSpreads && fragmentSpreads.length > 0) {
-      fragmentSpreads.forEach(fragmentSpread => {
-        fragmentsWithTypenameField[fragmentSpread] = true;
-      });
-    }
-
-    if (inlineFragments && inlineFragments.length > 0) {
-      const fragmentName = `${pascalCase(bareTypeName)}From${operationName}`;
-      handleInlineFragments(generator, fragmentName, inlineFragments);
-    }
-  });
 }
 
 export function typeDeclarationForFragment(
   generator,
-  {
+  fragment 
+) {
+  const {
     fragmentName,
     typeCondition,
     fields,
     inlineFragments,
     fragmentSpreads,
     source,
-    possibleTypes
-  },
-  fragmentsWithTypenameField
-) {
-  const interfaceName = `${pascalCase(fragmentName)}Fragment`;
+  } = fragment;
 
-  if (inlineFragments.length > 0) {
-    handleInlineFragments(generator, interfaceName, inlineFragments);
-  } else {
-    typeDeclaration(generator, {
-      interfaceName,
-      // extendTypes: fragmentSpreads ? fragmentSpreads.map(f => `${pascalCase(f)}Fragment`) : null,
-    }, () => {
-      if (fragmentsWithTypenameField[fragmentName]) {
-        addTypenameFieldIfNeeded(generator, fields, typeCondition);
-      }
+  const interfaceName = `${pascalCase(fragmentName)}Fragment`;
 
-      const properties = propertiesFromFields(generator.context, fields, {
-        typeNameSuffix: 'Fragment'
-      });
+  typeDeclaration(generator, {
+    interfaceName,
+    noBrackets: isAbstractType(typeCondition)
+  }, () => {
+    if (isAbstractType(typeCondition)) {
+      const propertySets = fragment.possibleTypes
+        .map(type => {
+          // NOTE: inlineFragment currently consists of the merged fields
+          // from both inline fragments and fragment spreads.
+          // TODO: Rename inlineFragments in the IR.
```

Comment:
@martijnwalraven Here is an IR change we can add for your next refactor. We'll not do it here because the PR is already huge.